rs/decentralization/src/network.rs

```diff
@@ -985,3 +985,85 @@ impl Display for SubnetChange {
         write!(f, "{}", SubnetChangeResponse::from(self))
     }
 }
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct NetworkHealSubnets {
+    pub name: String,
+    pub decentralized_subnet: DecentralizedSubnet,
+    pub unhealthy_nodes: Vec<Node>,
+}
+
+impl NetworkHealSubnets {
+    const IMPORTANT_SUBNETS: &'static [&'static str] = &["NNS", "SNS", "Bitcoin", "Internet Identity", "tECDSA signing"];
+}
+
+impl PartialOrd for NetworkHealSubnets {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for NetworkHealSubnets {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let self_is_important = NetworkHealSubnets::IMPORTANT_SUBNETS.contains(&self.name.as_str());
+        let other_is_important = NetworkHealSubnets::IMPORTANT_SUBNETS.contains(&other.name.as_str());
+
+        match (self_is_important, other_is_important) {
+            (true, false) => Ordering::Greater,
+            (false, true) => Ordering::Less,
+            _ => self.decentralized_subnet.nodes.len().cmp(&other.decentralized_subnet.nodes.len()),
+        }
+    }
+}
+
+pub struct NetworkHealRequest {
+    pub subnets: Vec<NetworkHealSubnets>,
+}
+
+impl NetworkHealRequest {
+    pub fn new(subnets: Vec<NetworkHealSubnets>) -> Self {
+        Self { subnets }
+    }
+
+    pub fn heal_and_optimize(
+        &self,
+        mut available_nodes: Vec<Node>,
+        max_replaceable_nodes: Option<usize>,
+    ) -> Result<Vec<SubnetChangeResponse>, NetworkError> {
+        let mut subnets_changed = Vec::new();
+        let subnets_to_heal = self.subnets.iter().sorted_by(|a, b| a.cmp(b).reverse()).collect_vec();
+
+        for subnet in subnets_to_heal {
+            let mut unhealthy_nodes = subnet.unhealthy_nodes.clone();
+            let unhealthy_nodes_len = unhealthy_nodes.len();
+
+            if let Some(max_replaceable_nodes) = max_replaceable_nodes {
+                if unhealthy_nodes_len > max_replaceable_nodes {
+                    unhealthy_nodes = subnet.unhealthy_nodes.clone().into_iter().take(max_replaceable_nodes).collect_vec();
+
+                    warn!(
+                        "Subnet {} has {} unhealthy nodes\nMax replacable nodes is {}\nNodes considered for replacement: {:?}",
+                        subnet.decentralized_subnet.id,
+                        unhealthy_nodes_len,
+                        max_replaceable_nodes,
+                        unhealthy_nodes.iter().map(|node| node.id).collect_vec()
+                    );
+                }
+            }
+            let unhealthy_nodes_len = unhealthy_nodes.len();
```

Comment:
Seems like a copy of line 1038

Reply:
This is necessary to update again the value